### PR TITLE
[FLINK-25796][network] Avoid record copy for result partition of sort-shuffle if there are enough buffers for better performance

### PR DESCRIPTION
**Pull request by wsry**
## What is the purpose of the change

Currently, for result partition of sort-shuffle, there is extra record copy overhead Introduced by clustering records by subpartition index. For small records, this overhead can cause even 20% performance regression. This ticket aims to solve the problem.

In fact, the hash-based implementation is a nature way to achieve the goal of sorting records by partition index. However, it incurs some serious weaknesses. For example, when there is no enough buffers or there is data skew, it can waste buffers and influence compression efficiency which can cause performance regression.

This ticket tries to solve the issue by dynamically switching between the two implementations, that is, if there are enough buffers, the hash-based implementation will be used and if there is no enough buffers, the sort-based implementation will be used.

## Brief change log

  - Dynamically switching between the two implementations, that is, if there are enough buffers, the hash-based implementation will be used and if there is no enough buffers, the sort-based implementation will be used.

## Verifying this change

This change added tests and existing tests can also help to verify the change.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/writer/ResultPartitionWriter.java [hunk 1]
```diff
@@ -41,7 +41,7 @@
  * ResultPartitionWriter#fail(Throwable)} still needs to be called afterwards to fully release all
  * resources associated the partition and propagate failure cause to the consumer if possible.
  */
-public interface ResultPartitionWriter extends AutoCloseable, AvailabilityProvider {
+public interface ResultPartitionWriter extends AvailabilityProvider {
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/writer/ResultPartitionWriter.java:44]: (One small comment as I'm not able to do full review): Why did you remove `AutoCloseable`?

Comment (wsry) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/writer/ResultPartitionWriter.java:44]: It is removed by accident. When I am checking where the close method is called, the IDE gives all places calling AutoCloseable#close, I removed it to reduce the result set size but forgot to add it back. Thanks for pointing it out. I will add it back.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SortMergeResultPartition.java [hunk 2]
```diff
@@ -103,6 +111,12 @@
      */
     private int numBuffersForSort;
 
+    /**
+     * Number of reserved network buffers for data writing. This value can be 0 and 0 means that
+     * {@link HashBasedPartitionSortedBuffer} will be used.
+     */
+    private int numBuffersForWrite;
```
Comment (gaoyunhaii) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SortMergeResultPartition.java:118]: From the following modification, it seems the variable here mainly plays a role to indicate whether we want to use hash-based implementation and sort-based implementation. I think perhaps we could directly use a variable like `sortBufferType` or `useHashBuffer` to make it more clear. The `numBuffersForWrite` could be changed to be a local variable in the constructor. We could change the implementation to be like 

```
if (numRequiredBuffer >= 2 * numSubpartitions) {
    useHashBuffer = true;
} else {
    useHashBuffer = false;
}

if (!useHashBuffer) {
    int expectedWriteBuffers;
    if (numRequiredBuffer >= 2 * numSubpartitions) {
        expectedWriteBuffers = 0;
    } else if (networkBufferSize >= NUM_WRITE_BUFFER_BYTES) {
        expectedWriteBuffers = 1;
    } else {
        expectedWriteBuffers =
            Math.min(EXPECTED_WRITE_BATCH_SIZE, NUM_WRITE_BUFFER_BYTES / networkBufferSize);
    }

    int numBuffersForWrite = Math.min(numRequiredBuffer / 2, expectedWriteBuffers);
    numBuffersForSort = numRequiredBuffer - numBuffersForWrite;

    try {
        for (int i = 0; i < numBuffersForWrite; ++i) {
            MemorySegment segment = bufferPool.requestMemorySegmentBlocking();
            writeSegments.add(segment);
        }
    } catch (InterruptedException exception) {
        // the setup method does not allow InterruptedException
        throw new IOException(exception);
    }
}
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SortMergeResultPartition.java [hunk 3]
```diff
@@ -288,58 +303,70 @@ private SortBuffer getBroadcastSortBuffer() throws IOException {
...
+        fileWriter.startNewRegion(isBroadcast);
+        do {
+            if (toWrite.size() >= numBuffersToWrite) {
+                writeBuffers(toWrite);
+                segments = new ArrayDeque<>(writeSegments);
             }
 
-            fileWriter.writeBuffers(toWrite);
-        }
+            BufferWithChannel bufferWithChannel = sortBuffer.copyIntoSegment(segments.poll());
```
Comment (gaoyunhaii) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SortMergeResultPartition.java:343]: It is also a bit weird that `copyIntoSegment` might pass a null segment. In consideration of the deadline I think we might rename the method to be like `getNextBuffer(@Nullable MemorySegment transitBuffer)` and add proper comments ?

**Review by gaoyunhaii — COMMENTED** [review]
Very thanks @wsry for the PR! I have left some comments. 

For the long run, I tend to we move the difference implementation to the implementation class, like the strategy to split write buffers and sort buffers and how they write the buffers to the files. Perhaps we could create a new issue for the future refactor? 

Also if possible I tend to we rename the classes to `DataBuffer`, `HashBasedDataBuffer` and `SortBasedDataBuffer` and also the variables to avoid reuse the word `sort`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SortMergeResultPartition.java [hunk 4]
```diff
@@ -365,16 +392,17 @@ private void updateStatistics(Buffer buffer, boolean isBroadcast) {
     private void writeLargeRecord(
             ByteBuffer record, int targetSubpartition, DataType dataType, boolean isBroadcast)
             throws IOException {
+        checkState(numBuffersForWrite > 0, "No buffers available for writing.");
```
[the remark this answers is not in the document]
Comment (wsry) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SortMergeResultPartition.java:395]: For hash-based implementation, large record will be appended to the sort buffer, when the data buffer is full the partial data of the record will be spilled as a data region and the remaining data of the large record will be appended to the sort buffer again. That is to say, a large record can span multiple data regions.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/HashBasedPartitionSortedBuffer.java [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+            writeRecord(source, targetChannel);
+        } else {
+            writeEvent(source, targetChannel, dataType);
+        }
+
+        isFull = source.hasRemaining();
+        if (!isFull) {
+            ++numTotalRecords;
+        }
+        numTotalBytes += totalBytes - source.remaining();
```
[the remark this answers is not in the document]
Comment (wsry) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/HashBasedPartitionSortedBuffer.java:148]: Yes, I will add some comments to explain that.

**Comment by wsry** [comment]
> Very thanks @wsry for the PR! I have left some comments.
> 
> For the long run, I tend to we move the difference implementation to the implementation class, like the strategy to split write buffers and sort buffers and how they write the buffers to the files. Perhaps we could create a new issue for the future refactor?
> 
> Also if possible I tend to we rename the classes to `DataBuffer`, `HashBasedDataBuffer` and `SortBasedDataBuffer` and also the variables to avoid reuse the word `sort`.

@gaoyunhaii Thanks for the review and comments. I agree that we can rename the sort buffer class in this PR. As for other refactor, I will create a new ticket to do it latter. I will update the PR soon.

**Comment by wsry** [comment]
@gaoyunhaii I have updated the PR.

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/SortMergeResultPartitionTest.java [hunk 6]
```diff
@@ -69,6 +72,8 @@
 
     private static final int numThreads = 4;
 
+    private final boolean useHashBasedSortBuffer;
```
Comment (gaoyunhaii) [inline flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/SortMergeResultPartitionTest.java:75]: nit: might change to `useHashBasedDataBuffer` ?

**Review by gaoyunhaii — APPROVED** [review]
LGTM % one small comment. 

Also @wsry could you squash the comments and rebase to the latest master to retrigger the ci pipeline? The architecture check issue should be fixed.

**Comment by wsry** [comment]
@flinkbot run azure
